lib/vfs/src/pipe.rs

```diff
@@ -0,0 +1,434 @@
+                        }
+                    }
+                }
+            };
+
+            rx.buffer.replace(Bytes::from(data));
+        }
+    }
+}
+
+impl VirtualFile for Pipe {
+    /// the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn last_accessed(&self) -> u64 {
+        0
+    }
+
+    /// the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn last_modified(&self) -> u64 {
+        0
+    }
+
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64 {
+        0
+    }
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64 {
+        0
+    }
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, _new_size: u64) -> Result<(), FsError> {
+        Ok(())
+    }
+
+    /// Request deletion of the file
+    fn unlink(&mut self) -> Result<(), FsError> {
+        Ok(())
+    }
+
+    /// Indicates if the file is opened or closed. This function must not block
+    /// Defaults to a status of being constantly open
+    fn is_open(&self) -> bool {
+        self.tx
+            .try_lock()
+            .map(|a| !a.is_closed())
+            .unwrap_or_else(|_| true)
+    }
+
+    /// Polls the file for when there is data to be read
+    fn poll_read_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<usize>> {
+        let mut rx = self.rx.lock().unwrap();
+        loop {
+            {
+                if let Some(inner_buf) = rx.buffer.as_mut() {
+                    let buf_len = inner_buf.len();
+                    if buf_len > 0 {
+                        return Poll::Ready(Ok(buf_len));
+                    }
+                }
+            }
+
+            let mut pinned_rx = Pin::new(&mut rx.chan);
+            let data = match pinned_rx.poll_recv(cx) {
+                Poll::Ready(Some(a)) => a,
+                Poll::Ready(None) => return Poll::Ready(Ok(0)),
+                Poll::Pending => {
+                    return match self.block {
+                        true => Poll::Pending,
+                        false => {
+                            Poll::Ready(Err(Into::<io::Error>::into(io::ErrorKind::WouldBlock)))
+                        }
+                    }
+                }
+            };
+
+            rx.buffer.replace(Bytes::from(data));
+        }
+    }
+
+    /// Polls the file for when it is available for writing
+    fn poll_write_ready(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<io::Result<usize>> {
+        let tx = self.tx.lock().unwrap();
+        if tx.is_closed() {
+            Poll::Ready(Ok(0))
+        } else {
+            Poll::Ready(Ok(8192))
+        }
+    }
+}
+
+/// Pipe pair of (a, b) Pipes that are connected together
+#[derive(Clone, Debug)]
+pub struct BidiPipe {
```

Comment:
Is this not normally called a stream duplex?

Reply:
I think Duplex does indeed sound better

Reply:
Yeah I agree, I renamed it to DuplexPipe. 

I stuck with pipe though because I think it fits better with our terminology, and it actually uses two pipes internally, so...